### PR TITLE
invidious: unstable-2023-06-06 -> unstable-2023-08-07

### PR DESCRIPTION
**Pull request by 999eagle**
## Description of changes

`hmac_key` has become a required configuration option for invidious, but as it is meant to be secret, it should be included using `services.invidious.extraSettingsFile` instead of adding a new option.

Changes are running on my invidious instance as of https://git.catgirl.cloud/999eagle/dotfiles-nix/-/commit/5bda4cb6805761b09a6bad70afd504c73feda73d.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by GaetanLepage — APPROVED** [review]
LGTM !

File: nixos/tests/invidious.nix
Comment (infinisil) [inline nixos/tests/invidious.nix:11]: How does the NixOS module behave when this isn't specified? Does it error at eval-time, build-time or run-time?

Comment (GaetanLepage) [reply nixos/tests/invidious.nix:11]: I would say that nix will not complain (at any stage) if `hmac_key` is not specified.
Maybe we can add an assertion to check if it is indeed provided by the user or not.

Comment (999eagle) [reply nixos/tests/invidious.nix:11]: Invidious will complain at run time if it's missing. There isn't really an easy way to check this though as `hmac_key` is supposed to be secret, so should be configured as raw json in `services.invidious.extraSettingsFile`.

Comment (GaetanLepage) [reply nixos/tests/invidious.nix:11]: Ok, then the only thing we could do is warn the user if neither `services.invidious.settings.hmac_key` or `services.invidious.extraSettingsFile` is not set.
But it might not even be our responsibility to make sure that someone's invidious config is correct or not.

Comment (infinisil) [reply nixos/tests/invidious.nix:11]: It's a really nice property for the service to just work by setting `enable = true`. Can the secret be generated at runtime? Can it change every time it starts or should it be persistent? If the latter it can be stored in `/var/lib/invidious` somewhere. And is there a reason a user would want to not use an auto-generated default?

Comment (GaetanLepage) [reply nixos/tests/invidious.nix:11]: yes I agree on the "work OOTB" argument.

[In the documentation](https://docs.invidious.io/installation/#post-install-configuration) it is stated that it can be generated randomly using the command `pwgen 20 1`.
[In the docker file](https://github.com/iv-org/documentation/pull/415/files), they expect the user to provide it.
Maybe it is fine if we generate it at build-time.

Comment (infinisil) [reply nixos/tests/invidious.nix:11]: Definitely shouldn't be done at build-time, otherwise the secret would end up in the Nix store. Should be run-time instead. If we had something like https://github.com/NixOS/nixpkgs/pull/179471 this would be easier to do, still possible without it though

Comment (miangraham) [reply nixos/tests/invidious.nix:11]: Any thoughts on how to add an hmac default without also breaking existing nixos module configs? Mine looks like this today:

```nix
invidious = {
  enable = true;
  package = unstable.invidious; # unstable-2023-07-05, already requires hmac_key
  settings = {
    # many things that are not hmac_key
  };
  extraSettingsFile = "/etc/invidious/config.yml"; # contains hmac_key
};
```

Obviously I can switch to doing whatever myself, but without checking extraSettingsFile for hmac_key's presence I'm not sure how you avoid stepping on toes for current users of unstable. Also not sure how big a concern it should be relative to friendliness to new people adding for the first time.

Comment (miangraham) [reply nixos/tests/invidious.nix:11]: The HMAC is used for app-level auth and CSRF tokens per https://github.com/iv-org/invidious/blob/master/config/config.example.yml#L437. I believe if it changes then you get some client credentials invalidated and errors thrown for existing sessions, users having to log out/clear cookies then log back in, etc. So you do want it to persist if possible, which is why they changed it to a mandatory config.

**Comment by miangraham** [comment]
Result of `nixpkgs-review pr 246447` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>invidious</li>
  </ul>
</details>

**Review by miangraham — APPROVED** [review]

File: nixos/tests/invidious.nix
Comment (infinisil) [reply nixos/tests/invidious.nix:11]: Oh I wasn't aware of `extraSettingsFile`, I didn't have that in my [original PR](https://github.com/NixOS/nixpkgs/pull/67664), and I don't think it should've been introduced, especially in light of NixOS/rfcs#42.

Instead it would've been best to have a `hmacKeyFile` option of type `nullOr path`, defaulting to `null`, implemented something like this:
```nix
{
  services.invidious.settings.hmac_key = lib.mkIf (cfg.hmacKeyFile != null) (lib.mkDefault "NIXOS_HMAC_KEY_AUTO_REPLACED");

  systemd.services.invidious.preStart =
    if cfg.hmacKeyFile != null then ''
      hmacKey=$(<${lib.escapeShellArg cfg.hmacKeyFile})
      sed "s/NIXOS_HMAC_KEY_AUTO_REPLACED/$hmacKey/" ${configFile} > config/config.yml
    '' else ''
      cp ${configFile} config/config.yml
    '';
}
```

Then for this update making requiring, `hmac_key` required, it could've easily been changed to this:
```nix
{
  services.invidious.settings.hmac_key = lib.mkDefault "NIXOS_HMAC_KEY_AUTO_REPLACED";

  systemd.services.invidious.preStart = ''
    ${if cfg.hmacKeyFile != null then ''
      hmacKeyFile=${lib.escapeShellArg cfg.hmacKeyFile}
    '' else ''
      hmacKeyFile=/var/lib/invidious/hmac_key
      if [[ ! -f "$hmacKeyFile" ]]; then
        # TODO: Ensure restrictive permissions
        pwgen 20 1 > "$hmacKeyFile"
      fi
    ''}
    hmacKey=$(<${lib.escapeShellArg cfg.hmacKeyFile})
    sed "s/NIXOS_HMAC_KEY_AUTO_REPLACED/$hmacKey/" ${configFile} > config/config.yml
  '';
}
```

Comment (infinisil) [reply nixos/tests/invidious.nix:11]: I would still be in favor of deprecating the `extraSettingsFile` option, mainly because it hurts the ability for other modules to inspect values via `settings`.

However, it turns out that `extraSettingsFile` is implemented nicely by merging it with the generated `settings` file using a `jq` filter: https://github.com/NixOS/nixpkgs/blob/c25960f7e89535622359790c03bcd9abfed7c298/nixos/modules/services/web-apps/invidious.nix#L19-L26

Which means we can easily extend the filter to merge it with another file. In particular such that `extraSettingsFile` takes precedence over `settings`, which takes precedence over an autogenerated config file with an `hmac_key`.
It could even be checked using `jq` whether `hmac_key` is already set anywhere before even generating the `hmac_key`!

**Review by infinisil — CHANGES_REQUESTED** [review]
The module should continue to Just Work™ using `enable = true`. We can do this by auto-generating an `hmac_key` if the user didn't provide one in the configuration, see https://github.com/NixOS/nixpkgs/pull/246447#discussion_r1288886992

**Comment by miangraham** [comment]
> The module should continue to Just Work™ using `enable = true`.

Just to be extra clear, this does not currently work on unstable and hasn't for a few months. This is due to the package-level runtime checks for the required setting added by upstream earlier this year. So this is a request for an improvement to the status quo on unstable, not an issue with this PR specifically.

Your call as to whether it's appropriate to block this merge on that improvement.

**Comment by infinisil** [comment]
@miangraham Shouldn't the NixOS test have prevented that? It only sets `enable = true`: https://github.com/NixOS/nixpkgs/blob/a378e68497eba05b76aa3790e8416fef6e7370d3/nixos/tests/invidious.nix#L9-L11

**Comment by miangraham** [comment]
@infinisil That would be a wonderful world. Here's the run from the last merge to unstable:
https://github.com/NixOS/nixpkgs/pull/243077/checks?check_run_id=14984517631

**Comment by infinisil** [comment]
@miangraham I'm not sure what you mean, that check seemed to run fine. The only notably line in the logs there relates to `hmac_key`, but the service didn't fail because of that
```
machine # [    8.709371] invidious-start[937]: 2023-07-12 15:07:37 UTC [warn]
  Please configure hmac_key by July 1st, see more here: https://github.com/iv-org/invidious/issues/3854
```

**Comment by miangraham** [comment]
@infinisil Diff from the same, no change to sha256: https://github.com/NixOS/nixpkgs/pull/243077/files

**Comment by infinisil** [comment]
@miangraham Oh damn, that's a problem! I opened https://github.com/NixOS/nixpkgs/pull/248186 to revert that

File: nixos/tests/invidious.nix
Comment (999eagle) [reply nixos/tests/invidious.nix:11]: I've got a mostly working version for this now. I'll test some more things and then push it, most likely tomorrow.

**Comment by miangraham** [comment]
Update for the record on why we were living in different realities here: unstable, for anyone not breaking on a hash mismatch, is runing stale code from unstable-2023-06-06 which is before the change to a mandatory hmac_key setting.

`services.invidious.enable = true;` by itself *does* in fact work for current versions on nixos/nixpkgs.

So this PR *is* in fact a good time to slow down and properly address the new config requirement.

Sorry for my part in the confusion here.

**Comment by 999eagle** [comment]
Rebased onto the current `nixpkgs-unstable` to resolve the merge conflict from #248186. `hmac_key` also works now.

**Comment by miangraham** [comment]
Tested upgrading with an existing extraSettingsFile hmac and everything runs fine here!

**Comment by miangraham** [comment]
Tested the latest package/module via overlay/disabledModules+imports and this looks great on my end. Works with or without my existing `extraSettingsFile`, which I can remove now in theory.

#### Verifying the new module effects:

New prestart entry exists
```console
$ grep ExecStart /etc/systemd/system/invidious.service
ExecStart=/nix/store/n01q5y7f6wzy1r1r3yph9lcsnp6kh212-unit-script-invidious-start/bin/invidious-start
ExecStartPre=/nix/store/saw8vnz0s3rljlv3phqbhmsgzsfdll8n-unit-script-invidious-pre-start/bin/invidious-pre-start
```

New auto-genned key exists
```console
$ wc -c /var/lib/invidious/hmac_key
21 /var/lib/invidious/hmac_key
```

New start command looks sane with existing `extraSettingsFile` on right-takes-precedence side of the merge:
```console
$ grep INVIDIOUS_CONFIG /nix/store/n01q5y7f6wzy1r1r3yph9lcsnp6kh212-unit-script-invidious-start/bin/invidious-start
export INVIDIOUS_CONFIG="$(/nix/store/i1jf9af43mf4jpnz7angp62bfndm8ksd-jq-1.6-bin/bin/jq -s ". | .[0].hmac_key = "'"'"$(cat '/var/lib/invidious/hmac_key')"'"'" | .[0] * .[1]" '/nix/store/vgm92w1lv3n0n226a2y2xbmkjjmjd79f-invidious-settings' '/etc/invidious/config.yml')"
```

And running and playback work without errors. Two thumbs up.

File: nixos/modules/services/web-apps/invidious.nix
Comment (infinisil) [inline nixos/modules/services/web-apps/invidious.nix:24]:
```suggestion
        if [[ ! -e "${finalHmacKeyFile}" ]]; then
```
The check for it being readable is not necessary, this file should always be readable.

File: nixos/modules/services/web-apps/invidious.nix
Comment (infinisil) [inline nixos/modules/services/web-apps/invidious.nix:37]: This is a real mess, if you have the time I'd really appreciate a refactoring of this. Ideally avoiding any escaping madness by having a constant `jq` expression and using `jq`'s `--arg` combined with `lib.escapeShellArg` to pass in values.

File: nixos/modules/services/web-apps/invidious.nix
Comment (infinisil) [inline nixos/modules/services/web-apps/invidious.nix:197]: This seems a bit wrong, defaults should never override explicit user settings. It should be this order from high to low:
- `hmacKeyFile`
- `extraSettingsFile`
- `settings` (though we should maybe complain about it not being safe if users specify the `hmac_key` here)
- An autogenerated default

I'm not sure if that's currently the case.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-apps/invidious.nix
Comment (999eagle) [reply nixos/modules/services/web-apps/invidious.nix:212]: I've changed the module to use this order now, keeping the `database.passwordFile` in between `settings` and `extraSettingsFile`.

File: nixos/modules/services/web-apps/invidious.nix
Comment (infinisil) [inline nixos/modules/services/web-apps/invidious.nix:212]:
```suggestion
        If non-`null`, this option overrides any `hmac_key` specified in {option}`services.invidious.settings`.
```

File: nixos/modules/services/web-apps/invidious.nix
Comment (infinisil) [inline nixos/modules/services/web-apps/invidious.nix:13]:
```suggestion
  generatedHmacKeyFile = "/var/lib/invidious/hmac_key";
```

This value is only ever used when `hmacKeyFile` is null.

File: nixos/modules/services/web-apps/invidious.nix
Comment (infinisil) [inline nixos/modules/services/web-apps/invidious.nix:43]:
```suggestion
        configParts+=("$(${pkgs.jq}/bin/jq -R '{"db":{"password":.}}' <${lib.escapeShellArg cfg.database.passwordFile})")
```

File: nixos/modules/services/web-apps/invidious.nix
Comment (infinisil) [inline nixos/modules/services/web-apps/invidious.nix:51]: These lines seem to have mismatch `"`'s, and you can always avoid `cat` here, and use `lib.escapeShellArgs` to have correct escaping:

```suggestion
        configParts+=("$(<${lib.escapeShellArg cfg.hmacKeyFile})")
```

File: nixos/modules/services/web-apps/invidious.nix
Comment (infinisil) [inline nixos/modules/services/web-apps/invidious.nix:47]:
```suggestion
        configParts+=("$(<${lib.escapeShellArg cfg.extraSettingsFile})")
```

File: nixos/modules/services/web-apps/invidious.nix
Comment (infinisil) [inline nixos/modules/services/web-apps/invidious.nix:39]: This doesn't strictly need `lib.escapeShellArg` because it's a Nix store path, which I'm pretty sure can't contain any sequence that would need to be escaped in bash, but let's do it anyways for consistency with my other suggestions, along with `cat`:

```suggestion
        configParts+=("$(<${lib.escapeShellArg settingsFile})")
```

**Review by infinisil — CHANGES_REQUESTED** [review]
This is looking pretty good! I have very minor suggestions still, but I think this should be good to merge after those are addressed :)

File: nixos/modules/services/web-apps/invidious.nix
Comment (999eagle) [reply nixos/modules/services/web-apps/invidious.nix]: Good catch, the quoting is indeed wrong. I do need `cat` though, because `jq` needs the file contents, not the file names in the final merge step.

Comment (infinisil) [reply nixos/modules/services/web-apps/invidious.nix]: You might have missed it or don't know it, but in Bash, `$(<foo)` is the same as `$(cat foo)` while avoiding an extra `cat`! I believe the code should work as I proposed it :)

**Review by infinisil — APPROVED** [review]
The PR looks good though, nice job with the `jq` refactoring! :100:

Regarding https://github.com/NixOS/nixpkgs/pull/246447#discussion_r1293849564, I don't really care much about that, so just ping me when you're happy with it, I'll merge it :wink:

**Comment by miangraham** [comment]
I've been running this since my last comment with zero issues. Just updated to https://github.com/NixOS/nixpkgs/commit/56eef8d0f96d11286cb1681efe953c1e1de3bea6 and everything still looks good. Ship it! 🚀

File: nixos/modules/services/web-apps/invidious.nix
Comment (999eagle) [reply nixos/modules/services/web-apps/invidious.nix]: Yeah that'd work too. The original suggestion didn't include the `<` though, so only the file path would be included. I've changed it to `<` now to stop the `cat` abuse :)

Comment (infinisil) [reply nixos/modules/services/web-apps/invidious.nix]: Oh weird, I can definitely see a `<` in my suggestion!

**Comment by infinisil** [comment]
Nice job @999eagle! And thanks for the testing @miangraham and @GaetanLepage! :rocket:

**Comment by GaetanLepage** [comment]
Thank you for getting this work done @infinisil @999eagle !

**Pull request merged**
